File: Payload_Type/odin/mythic/agent_functions/powershell_executor.py
from mythic_payloadtype_container.MythicCommandBase import *
import json
from mythic_payloadtype_container.MythicRPC import *


class PowerShellArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = []

    async def parse_arguments(self):
        pass


class PowerShellCommand(CommandBase):
    cmd = "powershell_executor"
    needs_admin = False
    help_cmd = "powershell_executor {command}"
    description = "Execute a shell command using 'powershell -nolog -noprofile'"
    version = 1
    author = "@antman1p"
    attackmapping = ["T1059", "T1059.001"]
    argument_class = PowerShellArguments
    attributes = CommandAttributes(
        suggested_command=True
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="powershell -nologo -noprofile {}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="{}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/freyja/freyja/agentfunctions/powershell_executor.go
package agentfunctions

import (
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

var powershell_executor = agentstructs.Command{
	Name:                      "powershell_executor",
  Author:                    "@antman1p",
	Description:               "Execute a shell command using 'powershell -nologo -noprofile'",
	MitreAttackMappings:       []string{"T1059"},
	CommandAttributes: agentstructs.CommandAttribute{
		CommandIsSuggested: true,
		CommandIsBuiltin: false,
		FilterCommandAvailabilityByAgentBuildParameters: true,
		SupportedOS: []string{agentstructs.SUPPORTED_OS_WINDOWS},
	},
	TaskFunctionCreateTasking: powershell_executorCreateTasking,
	TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
		return args.LoadArgsFromDictionary(input)
	},
	Version: 1,
}

func init() {
	agentstructs.AllPayloadData.Get("freyja").AddCommand(powershell_executor)
}

func powershell_executorCreateTasking(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
	response := agentstructs.PTTaskCreateTaskingMessageResponse{
		Success: true,
		TaskID:  taskData.Task.ID,
	}
	if _, err := mythicrpc.SendMythicRPCArtifactCreate(mythicrpc.MythicRPCArtifactCreateMessage{
		BaseArtifactType: "ProcessCreate",
		ArtifactMessage:  "powershell -nologo -noprofile " + taskData.Args.GetCommandLine(),
		TaskID:           taskData.Task.ID,
	}); err != nil {
		logging.LogError(err, "Failed to send mythicrpc artifact create")
	}
	return response
}
